File: app/src/main/java/io/catter2/favorites/GetFavoritesUseCase.java
package io.catter2.favorites;

import android.content.Context;

import java.util.ArrayList;
import java.util.List;

public class GetFavoritesUseCase {
    public interface Callback {
        void favoriteUrlsUpdated(List<String> favoriteUrls);
    }

    private FavoritesRepository repo;

    public GetFavoritesUseCase(Context context, String userToken) {
        this.repo = new FavoritesRepository(context, userToken);
    }

    /**
     * @param callback Callback returns a list of favorites once during registration and every time
     *                 the favorites are updated.
     */
    public void getFavorites(final Callback callback) {
        callback.favoriteUrlsUpdated(favoritesToUrls(repo.getFavorites()));

        repo.registerChangeListener(new FavoritesRepository.ChangeListener() {
            @Override
            public void onFavoritesChanged(List<FavoriteModel> favorites) {
                callback.favoriteUrlsUpdated(favoritesToUrls(favorites));
            }
        });
    }

    /**
     * Clear needs to be called when the use case if no more needed.
     */
    public void clear() {
        repo.clearChangeListener();
    }

    private List<String> favoritesToUrls(List<FavoriteModel> favorites) {
        ArrayList<String> urls = new ArrayList<>(favorites.size());
        for (FavoriteModel favorite : favorites) {
            urls.add(favorite.getUrl());
        }
        return urls;
    }
}

File: app/src/main/java/io/catter2/favorites/GetFavoritesUseCase.kt
package io.catter2.favorites

open class GetFavoritesUseCase(private val repo: FavoritesRepository) {
    /**
     * @param callback Callback returns a list of favorites once during registration and every time
     * the favorites are updated.
     */
    open fun getFavorites(callback: (List<String>) -> Unit) {
        callback(favoritesToUrls(repo.getFavorites()))
        repo.registerChangeListener(object : FavoritesRepository.ChangeListener {
            override fun onFavoritesChanged(favorites: List<FavoriteModel>) {
                callback(favoritesToUrls(favorites))
            }
        })
    }

    /**
     * Clear needs to be called when the use case is no more needed.
     */
    fun clear() {
        repo.clearChangeListener()
    }

    private fun favoritesToUrls(favorites: List<FavoriteModel>) = favorites.map { it.url }
}
